File: 2021/2021W25/PreppindataW25.py
import pandas as pd
import re

xlsx = pd.ExcelFile("F:/Data/2021W25 Input.xlsx")
gen1 = pd.read_excel(xlsx,sheet_name='Gen 1')[['#','Name']].dropna()

evogroup = pd.read_excel(xlsx,sheet_name='Evolution Group')
evogroup['#']=evogroup['#'].apply(lambda x: int(x.strip()))
evogroup = evogroup[(evogroup['Starter?']==0) & (evogroup['Legendary?']==0)][['#','Evolution Group']]

gen1g = pd.merge(gen1, evogroup, on='#').drop('#', axis=1)

evo = pd.read_excel(xlsx,sheet_name='Evolutions')[['Evolving from', 'Evolving to']]

df = pd.concat([
    pd.merge(evo, gen1g, left_on='Evolving from', right_on='Name')[['Evolution Group','Evolving to']].rename(
        columns={'Evolving to':'Pokemon'}),
    pd.merge(evo, gen1g, left_on='Evolving to', right_on='Name')[['Evolution Group','Evolving from']].rename(
        columns={'Evolving from':'Pokemon'})]).drop_duplicates()

mega = pd.concat([pd.read_excel(xlsx,sheet) for sheet in ['Mega Evolutions', 'Alolan', 'Galarian', 'Gigantamax']])
mega['Name'] = mega['Name'].apply(lambda x: re.sub('^\w+\s','',x))
mega = mega.drop_duplicates()

df['Check'] = df['Pokemon'].isin(list(mega['Name'])).astype(int)
df['Group Check'] = df['Check'].groupby(df['Evolution Group']).transform('sum') 
df = df[df['Group Check']==0][['Evolution Group', 'Pokemon']].copy()

unatt = pd.read_excel(xlsx,sheet_name='Unattainable in Sword & Shield').rename(columns={'Name':'Evolution Group'})
anime = pd.read_excel(xlsx,sheet_name='Anime Appearances')

df = pd.merge(df, unatt, on='Evolution Group')
df = pd.merge(df, anime, on='Pokemon')
df = df.groupby('Evolution Group', as_index=False).agg(Appearances=('Episode','nunique'))
df['The Worst Pokemon'] = df['Appearances'].rank(method='min').astype(int)
df = df[['The Worst Pokemon', 'Evolution Group', 'Appearances']].sort_values('Appearances')

File: 2021/2021W25/preppindataW25.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)

xlsx <- "F:/Data/2021W25 Input.xlsx"

grp <- read_excel(xlsx, sheet = 'Evolution Group') %>%
  filter(`Starter?`==0 & `Legendary?`==0) %>%
  select(c('#','Evolution Group')) %>%
  mutate('#' = as.integer(str_remove(`#`,'\\D')))

evo <- read_excel(xlsx, sheet = 'Evolutions') %>% 
  select(c('Evolving from', 'Evolving to'))

gen1 <- read_excel(xlsx, sheet = 'Gen 1') %>%
  select(c('#','Name')) %>%
  drop_na() %>%
  merge(grp, by='#') 

mega <- xlsx %>%
  excel_sheets() %>% .[grepl('(Mega Evolutions|Alolan|Galarian|Gigantamax)',.)] %>%
  set_names() %>%
  map_df( ~ read_excel(xlsx, sheet = .x,)) %>%
  mutate('Name' = str_remove(`Name`, '^(\\w+\\s)')) %>%
  distinct()

final <- rbind(merge(evo, gen1, by.x='Evolving from', by.y='Name') %>%
                 rename('Pokemon' = 'Evolving to') %>%
                 select(c('Evolution Group', 'Pokemon')),
               merge(evo, gen1, by.x='Evolving to', by.y='Name') %>%
                 rename('Pokemon' = 'Evolving from') %>%
                 select(c('Evolution Group', 'Pokemon'))) %>%
  distinct() %>%
  mutate('Check' = as.integer(`Pokemon` %in% mega$Name)) %>%
  group_by(`Evolution Group`) %>%
  mutate('Group Check' = sum(`Check`)) %>%
  filter(`Group Check`==0) %>%
  merge(read_excel(xlsx, sheet = 'Unattainable in Sword & Shield') %>%
          rename('Evolution Group'='Name'), by='Evolution Group') %>%
  merge(read_excel(xlsx, sheet = 'Anime Appearances'), by='Pokemon') %>%
  group_by(`Evolution Group`) %>%
  summarise('Appearances' = n_distinct(`Episode`), .groups='drop') %>%
  mutate('The Worst Pokemon' = as.integer(rank(`Appearances`, ties.method='min'))) %>%
  select(c('The Worst Pokemon', 'Evolution Group', 'Appearances')) %>%
  arrange(`Appearances`) 

View(final)
